flocker/node/agents/blockdevice.py

```diff
@@ -527,23 +553,28 @@ class AttachVolume(PRecord):
     :ivar unicode hostname: An identifier for the node to which the volume
         should be attached.  An IPv4 address literal.
     """
-    volume = _volume_field()
+    dataset_id = field(type=UUID, mandatory=True)
     hostname = field(type=unicode, mandatory=True)
 
     @property
     def eliot_action(self):
-        return ATTACH_VOLUME(_logger, volume=self.volume)
+        return ATTACH_VOLUME(_logger, dataset_id=self.dataset_id)
 
     def run(self, deployer):
         """
         Use the deployer's ``IBlockDeviceAPI`` to attach the volume.
         """
-        # Make this asynchronous after FLOC-1549, probably as part of
-        # FLOC-1593.
-        deployer.block_device_api.attach_volume(
-            self.volume.blockdevice_id, self.hostname
+        api = deployer.async_block_device_api
+        listing = api.list_volumes()
+        listing.addCallback(
+            _blockdevice_volume_from_datasetid, self.dataset_id
         )
-        return succeed(None)
+
+        def found(volume):
+            ATTACH_VOLUME_DETAILS(volume=volume).write(_logger)
```

Comment:
And coverage shows that that case never gets tested:
 * file:///home/richard/projects/HybridLogic/flocker/htmlcov/flocker_node_agents_blockdevice.html#n288

Is it possible that a block device might have been destroyed between the calculation of the state changes and the running of the change? It sounds plausible.